src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights/ApplicationInsightsEventPublisher.cs

```diff
@@ -0,0 +1,60 @@
+﻿using Microsoft.ApplicationInsights;
+using System.Diagnostics;
+
+namespace Microsoft.FeatureManagement.Telemetry.ApplicationInsights
+{
+    /// <summary>
+    /// Listens to <see cref="Activity"/> events from feature management and sends them to Application Insights.
+    /// </summary>
+    internal sealed class ApplicationInsightsEventPublisher : IDisposable
+    {
+        private readonly TelemetryClient _telemetryClient;
+        private readonly ActivityListener _activityListener;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ApplicationInsightsEventPublisher"/> class.
+        /// </summary>
+        /// <param name="telemetryClient">The Application Insights telemetry client.</param>
+        public ApplicationInsightsEventPublisher(TelemetryClient telemetryClient)
+        {
+            _telemetryClient = telemetryClient ?? throw new ArgumentNullException(nameof(telemetryClient));
+
+            _activityListener = new ActivityListener
+            {
+                ShouldListenTo = (activitySource) => activitySource.Name == "Microsoft.FeatureManagement",
+                Sample = (ref ActivityCreationOptions<ActivityContext> options) => ActivitySamplingResult.AllData,
+                ActivityStopped = (activity) =>
+                {
+                    ActivityEvent? evaluationEvent = activity.Events.FirstOrDefault((activityEvent) => activityEvent.Name == "feature_flag");
+
+                    if (evaluationEvent != null && evaluationEvent.Value.Tags.Any())
+                    {
+                        HandleActivityEvent(evaluationEvent.Value);
+                    }
+                }
+            };
+
+            ActivitySource.AddActivityListener(_activityListener);
+        }
+
+        /// <summary>
+        /// Disposes the resources used by the <see cref="ApplicationInsightsEventPublisher"/>.
+        /// </summary>
+        public void Dispose()
+        {
+            _activityListener.Dispose();
+        }
+
+        private void HandleActivityEvent(ActivityEvent activityEvent)
+        {
+            var properties = new Dictionary<string, string>();
+
+            foreach (var tag in activityEvent.Tags)
+            {
+                properties[tag.Key] = tag.Value?.ToString();
```

Comment:
Do you think it's necessary for us to add some additional checks here to ensure that certain fields we need internally (for example "VariantAssignmentReason") will definitely appear in the app insights event? Currently, if the activity event doesn't have these tags, then the property of the app insights event won't have these fields.

Reply:
I think no. In the case where the event isn't correct, we should still emit the event because it's still telemetry and this is the publisher for telemetry. For example, if FM is used without the provider, etag won't exist which is required for experimentation, but not required for telemetry.